# Part 2: Weather -> Daily Electricity Consumption Prediction

**Dataset:** Electricity Consumption Based on Weather Data (Kaggle- https://www.kaggle.com/datasets/sudhirsingh27/electricity-consumption-based-on-weather-data)

**Task:** Build regression models to predict daily electricity consumption using weather features.

**Models:**
- Linear Regression
- Polynomial Regression (degree 2, 3, 4)

## 2.1 Load and Inspect the Dataset

In [17]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [ ]:
# Load the dataset
df = pd.read_csv("data/electricity_consumption_based_weather_dataset.csv")

# Display column names and shape
print("Column Names:")
print(df.columns.tolist())
print("\nShape (rows, columns):", df.shape)

Column Names:
['date', 'AWND', 'PRCP', 'TMAX', 'TMIN', 'daily_consumption']

Shape (rows, columns): (1433, 6)


In [19]:
# Display first few rows
print("First 5 rows:")
df.head()

First 5 rows:


,date,AWND,PRCP,TMAX,TMIN,daily_consumption
0,2006-12-16,2.5,0.0,10.6,5.0,1209.176
1,2006-12-17,2.6,0.0,13.3,5.6,3390.460
2,2006-12-18,2.4,0.0,15.0,6.7,2203.826
3,2006-12-19,2.4,0.0,7.2,2.2,1666.194
4,2006-12-20,2.4,0.0,7.2,1.1,2225.748


In [20]:
# Summary statistics
print("Summary Statistics:")
df.describe()

Summary Statistics:


,AWND,PRCP,TMAX,TMIN,daily_consumption
count,1418.000000,1433.000000,1433.000000,1433.000000,1433.000000
mean,2.642313,3.800488,17.187509,9.141242,1561.078061
std,1.140021,10.973436,10.136415,9.028417,606.819667
min,0.000000,0.000000,-8.900000,-14.400000,14.218000
25%,1.800000,0.000000,8.900000,2.200000,1165.700000
50%,2.400000,0.000000,17.800000,9.400000,1542.650000
75%,3.300000,1.300000,26.100000,17.200000,1893.608000
max,10.200000,192.300000,39.400000,27.200000,4773.386000


In [21]:
# Check for missing values
print("Missing Values per Column:")
print(df.isna().sum())
print(f"\nTotal Missing Values: {df.isna().sum().sum()}")

Missing Values per Column:
date                  0
AWND                 15
PRCP                  0
TMAX                  0
TMIN                  0
daily_consumption     0
dtype: int64

Total Missing Values: 15


In [22]:
# Handle missing values by dropping rows with any missing data
df_clean = df.dropna()

print(f"Original shape: {df.shape}")
print(f"After dropping missing values: {df_clean.shape}")
print(f"Rows removed: {df.shape[0] - df_clean.shape[0]}")

Original shape: (1433, 6)
After dropping missing values: (1418, 6)
Rows removed: 15


### Dataset Summary

**Target (Dependent Variable):** `daily_consumption` — daily electricity consumption

**Features (Independent Variables):**
- `AWND` — Average wind speed
- `PRCP` — Precipitation
- `TMAX` — Maximum temperature
- `TMIN` — Minimum temperature

**Handling Missing Values:** Rows with missing values are dropped using `dropna()`.

## 2.2 Train/Test Split (70% / 30% Random)

In [23]:
# Define features (X) and target (y)
# Features: weather variables (AWND, PRCP, TMAX, TMIN)
# Target: daily_consumption

X = df_clean[["AWND", "PRCP", "TMAX", "TMIN"]]
y = df_clean["daily_consumption"]

print("Features (X) shape:", X.shape)
print("Target (y) shape:", y.shape)

Features (X) shape: (1418, 4)
Target (y) shape: (1418,)


In [24]:
# Split into train (70%) and test (30%)
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.30,
    random_state=42
)

print("Training set size:", len(X_train))
print("Testing set size:", len(X_test))
print(f"\nTrain/Test ratio: {len(X_train)/len(X)*100:.0f}% / {len(X_test)/len(X)*100:.0f}%")

Training set size: 992
Testing set size: 426

Train/Test ratio: 70% / 30%


## 2.3 Model Training: Linear + Polynomial Regression

1. **Linear Regression**
2. **Polynomial Regression (degree 2, 3, 4)**

No regularization used.

In [25]:
# Model 1: Linear Regression
model_linear = LinearRegression()
model_linear.fit(X_train, y_train)

print("Linear Regression trained.")

Linear Regression trained.


In [26]:
# Model 2: Polynomial Regression (Degree 2)
model_poly2 = Pipeline([
    ("poly", PolynomialFeatures(degree=2, include_bias=False)),
    ("lr", LinearRegression())
])
model_poly2.fit(X_train, y_train)

print("Polynomial Regression (degree=2) trained.")

Polynomial Regression (degree=2) trained.


In [27]:
# Model 3: Polynomial Regression (Degree 3)
model_poly3 = Pipeline([
    ("poly", PolynomialFeatures(degree=3, include_bias=False)),
    ("lr", LinearRegression())
])
model_poly3.fit(X_train, y_train)

print("Polynomial Regression (degree=3) trained.")

Polynomial Regression (degree=3) trained.


In [28]:
# Model 4: Polynomial Regression (Degree 4)
model_poly4 = Pipeline([
    ("poly", PolynomialFeatures(degree=4, include_bias=False)),
    ("lr", LinearRegression())
])
model_poly4.fit(X_train, y_train)

print("Polynomial Regression (degree=4) trained.")

Polynomial Regression (degree=4) trained.


## 2.4 Model Evaluation (Train and Test)

Metrics: **MSE**, **MAE**, **R²** on both train and test sets.

In [29]:
# Helper function to compute metrics
def compute_metrics(y_true, y_pred):
    return {
        "MSE": mean_squared_error(y_true, y_pred),
        "MAE": mean_absolute_error(y_true, y_pred),
        "R^2": r2_score(y_true, y_pred),
    }

In [30]:
# Evaluate all models
models = [
    ("Linear Regression", model_linear),
    ("Polynomial (degree=2)", model_poly2),
    ("Polynomial (degree=3)", model_poly3),
    ("Polynomial (degree=4)", model_poly4),
]

results = []
for model_name, model in models:
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    train_metrics = compute_metrics(y_train, y_train_pred)
    test_metrics = compute_metrics(y_test, y_test_pred)

    results.append({
        "Model": model_name,
        "Train MSE": train_metrics["MSE"],
        "Train MAE": train_metrics["MAE"],
        "Train R²": train_metrics["R^2"],
        "Test MSE": test_metrics["MSE"],
        "Test MAE": test_metrics["MAE"],
        "Test R²": test_metrics["R^2"],
    })

In [31]:
# Display results table
results_df = pd.DataFrame(results)

print("=" * 80)
print("Model Comparison: Train vs Test Metrics")
print("=" * 80)
results_df

Model Comparison: Train vs Test Metrics


,Model,Train MSE,Train MAE,Train R²,Test MSE,Test MAE,Test R²
0,Linear Regression,272403.396174,384.465016,0.276000,2.481258e+05,375.404537,0.299333
1,Polynomial (degree=2),264765.769932,379.648753,0.296300,2.552685e+05,379.039083,0.279163
2,Polynomial (degree=3),259249.534870,375.952901,0.310961,2.656237e+05,385.235167,0.249922
3,Polynomial (degree=4),251909.339001,372.116566,0.330470,1.215149e+07,578.642201,-33.313844


## 2.5 Discussion and Interpretation

### Which model generalizes best?

**Linear Regression** generalizes best with the highest Test R² (0.299) and lowest Test MSE (248,126).

This suggests the relationship between weather and electricity consumption is approximately linear — adding polynomial terms does not capture meaningful patterns, only noise.

### Do polynomial models improve the fit?

**No.** Polynomial models perform worse on the test set:
- Linear → Poly(2): Test R² decreased from 0.299 to 0.279
- Poly(2) → Poly(3): Test R² decreased from 0.279 to 0.250
- Poly(3) → Poly(4): Test R² **collapsed to -33.31** (catastrophic overfitting)

While Train R² increases with degree (0.276 → 0.296 → 0.311 → 0.330), this improvement doesn't generalize — the models are fitting training noise.

### Why does Polynomial (degree=4) fail so badly?

With 4 features and degree 4, `PolynomialFeatures` creates **many interaction terms** (70 features for degree 4). This leads to:
1. **Severe overfitting:** The model memorizes training data patterns that don't exist in test data
2. **Unstable coefficients:** Too many parameters for the signal available
3. Evidence: Train R² = 0.330 but Test R² = -33.31 (worse than predicting the mean)

### Why is overall model performance low (R² ≈ 0.30)?

All models achieve relatively low R² (~0.25-0.30). Two likely reasons:

1. **Limited feature set:** Weather alone (wind, precipitation, temperature) doesn't fully explain electricity consumption. Missing factors include:
   - Day of week / holidays (behavioral patterns)
   - Occupancy and human activity
   - Time of year / seasonal effects beyond temperature

2. **High noise / variability:** Daily consumption has significant variation (std = 607) not explained by weather, likely driven by unpredictable human behavior.

### Summary

| Finding | Evidence |
|---------|----------|
| Best model: Linear Regression | Highest Test R² (0.299) |
| Polynomial models hurt performance | Test R² decreases with higher degree |
| Degree 4 catastrophically overfits | Test R² = -33.31 |
| Weather explains ~30% of variance | All models R² ≈ 0.25-0.30 |

**Conclusion:** For this dataset, Linear Regression is the best choice. The weather features provide limited predictive power for electricity consumption — additional features (time, occupancy, etc.) would likely be needed to improve model performance.